src/reduced_lung/src/4C_reduced_lung_aaa_approx.cpp

```diff
@@ -0,0 +1,287 @@
+  {
+    int M = Z.size();
+
+    // Evaluate target function at support points
+    const std::vector<double> F = target_function(Z);
+
+    // J = candidate indices of support points
+    std::vector<int> J(M);
+    std::iota(J.begin(), J.end(), 0);
+
+    // Output vectors of barycentric approximation
+    std::vector<double> z, f, w, errvec;
+    // Cauchy matrix (iteratively filled)
+    Core::LinAlg::SerialDenseMatrix C(M, opts.mmax, true);
+
+    // R = initial approximant = mean(F)
+    std::vector<double> R(M);
+    const double f_mean = std::accumulate(F.begin(), F.end(), 0.0) / M;
+    const double f_max = *std::ranges::max_element(F);
+    std::ranges::fill(R, f_mean);
+
+    // Main loop with greedy selection of support points
+    for (int m = 0; m < opts.mmax; ++m)
+    {
+      int j = -1;
+      double max_err = 0.0;
+      // Loop over all remaining support point indices
+      for (int i : J)
+      {
+        // Calculate maximal approximation error
+        double diff = std::abs(F[i] - R[i]);
+        if (diff >= max_err)
+        {
+          max_err = diff;
+          j = i;
+        }
+      }
+
+      // Add support point with maximum error
+      z.push_back(Z[j]);
+      f.push_back(F[j]);
+
+      // Remove j from J
+      std::erase(J, j);
+
+      // update C(:,m)
+      for (int i = 0; i < M; ++i)
+      {
+        if (i == j)
+        {
+          C(i, m) = 0.0;  // unused row, avoid division by zero
+        }
+        else
+        {
+          C(i, m) = 1.0 / (Z[i] - Z[j]);
+        }
+      }
+
+      // build Loewner matrix A(J,:)
+      int n = static_cast<int>(J.size());
+      Core::LinAlg::SerialDenseMatrix A(n, m + 1);
+      for (int row = 0; row < n; ++row)
+      {
+        int i = J[row];
+        for (int col = 0; col <= m; ++col) A(row, col) = F[i] * C(i, col) - C(i, col) * f[col];
+      }
+
+      // SVD of A
+      Core::LinAlg::SerialDenseMatrix U(n, n, true);
+      Core::LinAlg::SerialDenseMatrix S(n, m + 1, true);
+      Core::LinAlg::SerialDenseMatrix Vt(m + 1, m + 1, true);
+      Core::LinAlg::svd(A, U, S, Vt);
+
+      // w = last column of V
+      w.resize(m + 1);
+      for (int k = 0; k <= m; ++k) w[k] = Vt(m, k);
+
+      // Compute new approximation R(J) = N(J)/D(J)
+      for (int idx : J)
+      {
+        double num = 0.0, den = 0.0;
+        for (int k = 0; k <= m; ++k)
+        {
+          num += C(idx, k) * w[k] * f[k];
+          den += C(idx, k) * w[k];
+        }
+        R[idx] = num / den;
+      }
+      R[j] = F[j];
+
+      // Calculate new approximation error
+      double err = 0.0;
+      for (int i = 0; i < M; ++i) err = std::max(err, std::abs((F[i] - R[i]) / f_max));
+      errvec.push_back(err);
+      if (err <= opts.tol) break;
+    }
+
+    AAAResult result;
+    result.z = std::move(z);
+    result.f = std::move(f);
+    result.w = std::move(w);
+    result.errvec = std::move(errvec);
```

Comment:
You could also create the `result` at the top and fill the values in there during the algorithm.